src/utils/DataSourceHelper/DataSourceHelper.ts

```diff
@@ -0,0 +1,91 @@
+import "./ui/DataSourceHelper.scss";
+
+interface ConstraintStore {
+    [widgetId: string]: string | HybridConstraint;
+}
+
+interface Version {
+    major: number; minor: number; path: number;
+}
+
+interface HybridConstraint {
+    attribute: string;
+    operator: string;
+    value: string;
+    path?: string;
+}
+
+interface ListView extends mxui.widget._WidgetBase {
+    _datasource: {
+        _constraints: HybridConstraint[] | string;
+        _pageObjs: mendix.lib.MxObject[];
+    };
+    datasource: {
+        type: "microflow" | "entityPath" | "database" | "xpath";
+    };
+    update: (obj: mendix.lib.MxObject | null, callback?: () => void) => void;
+    _entity: string;
+}
+
+export class DataSourceHelper {
+    static VERSION: Version = { major: 1, minor: 0, path: 0 };
+    version: Version;
+    private delay = 100; // TODO check optimal timing.
+    private timeoutHandle?: number;
+    private store: ConstraintStore;
+    private widget: ListView;
+
+    constructor(widget: ListView) {
+        this.widget = widget;
+
+        this.compatibilityCheck();
+
+        this.store = {};
+        this.version = DataSourceHelper.VERSION;
+    }
+
+    setConstraint(widgetId: string, constraint: string | HybridConstraint) {
+        this.store[widgetId] = constraint;
+        if (this.timeoutHandle) {
+            window.clearTimeout(this.timeoutHandle);
+        }
+        this.timeoutHandle = window.setTimeout(() => {
+            this.applyConstraint();
+        }, this.delay);
+    }
+
+    public static checkVersionCompatible(version: Version): boolean {
+        return DataSourceHelper.VERSION.major === version.major;
+    }
+
+    private compatibilityCheck() {
+        if (!(this.widget._datasource && this.widget._datasource._constraints !== undefined && this.widget._entity && this.widget.update)) {
```

Comment:
Why `this.widget._datasource._constraints !== undefined ` ?

Reply:
when `this.widget._datasource._constraints` is `""`
`if(this.widget._datasource._constraints)` will always be `false` hence not compatible. Yet we are checking if `this.widget._datasource._constraints` is defined/exists on the widget. For that reason we use negated `this.widget._datasource._constraints !== undefined`